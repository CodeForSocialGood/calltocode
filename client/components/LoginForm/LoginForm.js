import React, {Component} from 'react'
import {connect} from 'react-redux'
import {Link} from 'react-router-dom'
import PropTypes from 'prop-types'

import AuthActionCreator from '../../actions/auth'
import styles from './LoginForm.scss'
import {buttonSubmit} from './loginFormJss'
import {withStyles} from 'material-ui/styles'
import HeaderActionCreator from '../../actions/header'

/**
 * material ui components
 */
import TextField from 'material-ui/TextField'
import Button from 'material-ui/Button'

class LoginForm extends Component {
  constructor (props, context) {
    super(props, context)
    this.state = {
      email: '',
      password: '',
      error: {},
      isValid: false
    }
    this.handleChange = this.handleChange.bind(this)
    this.onBlur = this.onBlur.bind(this)
    this.onSubmit = this.onSubmit.bind(this)
  }

  async onSubmit (event) {
    event.preventDefault()
    if (this.state.email.length === 0) {
      this.setState({
        error: {
          ...this.state.error,
          'email': this.state.email.length === 0,
          'password': this.state.password.length === 0
        }
      })
    }
    if (this.state.password.length === 0 || this.state.email.length === 0) {
      return
    }
    const loginRes = await this.props.doLogin(this.state.email, this.state.password)
    if (loginRes) {
      this.context.router.history.push('/')
    }
  }

  onBlur (event) {
    event.preventDefault()
    this.setState({
      error: {...this.state.error, [event.target.name]: this.state[event.target.name].length === 0}
    })
  }

  handleChange (event) {
    event.preventDefault()
    this.setState({[event.target.name]: event.target.value})
  }

  enableOrDisableLogin () {
    const email = this.state.email || ''
    const password = this.state.password || ''
    const {enableLogin, disableLogin} = this.props

    if (!email.trim() || !password.trim()) {
      disableLogin()
    } else {
      enableLogin()
    }
  }

  componentWillUnmount () {
    const {enableLogin} = this.props
    enableLogin()
  }

  render() {
    const {classes} = this.props

    this.enableOrDisableLogin()

    return (
      <form id="loginForm" className={styles.form} onSubmit={this.onSubmit}>
        <h1 className={styles.h1}>Login</h1>

        <TextField required id="email" error={this.state.error['email']} label="Email" type="text" fullWidth
                   className={styles.inputEmail} name="email"
                   onChange={this.handleChange} onBlur={this.onBlur}/>

        <TextField required id="password" error={this.state.error['password']} label="Password" type="password"
                   fullWidth className={styles.inputPassword} name="password"
                   onChange={this.handleChange} onBlur={this.onBlur}/>

        <Button type="submit" raised className={classes.root} color="primary" fullWidth={true}>
          Submit
        </Button>

        <div className={styles.errorContent}>
          {this.props.error}
        </div>

        <Link
          className={styles.forgotPassword}
          key='forgotPassword'
          to='/forgot-password'>
          Forgot Password?
        </Link>
      </form>
    )
  }
}

const mapDispatchToProps = {
  doLogin: AuthActionCreator.doLogin,
  enableLogin: HeaderActionCreator.enableLogin,
  disableLogin: HeaderActionCreator.disableLogin
};

const mapStateToProps = (state) => {
  return {
    error: state.auth.error
  }
}

LoginForm.contextTypes = {
  router: PropTypes.object
}

LoginForm.propTypes = {
  error: PropTypes.string,
  handleSubmit: PropTypes.func,
  doLogin: PropTypes.func,
  classes: PropTypes.object,
  enableLogin: PropTypes.func,
  disableLogin: PropTypes.func
}

export default connect(mapStateToProps, mapDispatchToProps)(withStyles(buttonSubmit)(LoginForm))
