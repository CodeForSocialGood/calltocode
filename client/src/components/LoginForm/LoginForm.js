import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Field, reduxForm, SubmissionError } from 'redux-form'
import { push } from 'react-router-redux'
import PropTypes from 'prop-types'

import styles from './LoginForm.scss'
import { login } from '../../actions'
import loginApiClient from '../../api/login'

class LoginForm extends Component {
  renderEmail (field) {
    return (
      <input className={styles.inputEmail}
        placeholder="Email"
        {...field.input} />
    )
  }

  renderPassword (field) {
    return (
      <input className={styles.inputPassword}
        placeholder="Password"
        type="password"
        {...field.input} />
    )
  }

  async validateEmailAndPassword (values) {
    const { email, password } = values
    const response = await loginApiClient.login(email, password)

    if (response.status === 403) {
      throw new SubmissionError({ email, _error: 'Unauthorized!' })
    } else {
      this.props.login({ email })
    }
  }

  render () {
    const { handleSubmit } = this.props

    return (
      <form className={styles.form} onSubmit={handleSubmit(this.validateEmailAndPassword.bind(this))}>
        <h1 className={styles.title}>Login</h1>

        <Field name="email"
          component={this.renderEmail} />
        <Field name="password"
          component={this.renderPassword} />

        <button className={styles.buttonSubmit} type="submit">
          Submit
        </button>
      </form>
    )
  }
}

LoginForm.propTypes = {
  login: PropTypes.func,
  handleSubmit: PropTypes.func
}

const LoginFormRedux = reduxForm({
  form: 'LoginForm',
  onSubmitSuccess: (result, dispatch) => dispatch(push('/'))
})(LoginForm)

export default connect(null, { login })(LoginFormRedux)
