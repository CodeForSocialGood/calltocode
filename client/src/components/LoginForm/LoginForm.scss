@import '../../scss/partials/partials';
@import '../../scss/elements/form';

.form {
  @extend %form;
}

h1 {
  @extend %formTitle;
  grid-column: 1 / 1;
  grid-row: 1;
}

.inputEmail {
  @extend %fieldText;
  grid-column: 1 / 1;
  grid-row: 2;
}

.inputPassword {
  @extend %fieldText;
  grid-column: 1 / 1;
  grid-row: 3;
}

.buttonSubmit {
  @extend %buttonSubmit;
  grid-column: 1 / 1;
  grid-row: 4;
}

.buttonError {
  @extend %button;
  background-color: $errorBtnColor;
  padding: 20px;
  grid-column: 1 / 1;
  grid-row: 4;
}

.errorContent {
  grid-row: 5;
  grid-column: 1 / 1;
  font-size: 1rem;
  color: $errorColor;
}

.forgotPassword {
  grid-column: 1 / 1;
  grid-row: 6;
  margin: 20px auto;
  text-align: center;
  color: $primaryColor;

  &:hover {
    color: $primaryColorHover;
  }
}

.tick:before{
  content: "\2713 ";
  color: darkgreen;
}

.cross:before{
  content: "\2717 ";
  color: crimson;
}

.show{
  display: block;
}

.hide{
  display: none;
}

.validpopup{
  background-color: #fff;
  max-width: 25vw;
  font-size: .8em;
  padding: 3%;
  grid-area: 4 / 1 / span 4 / 1 ;
  z-index: 1;
  left: 10%;
  position: relative;
  li {
    margin-top: 3%;
  }
  @extend %boxShadowSmall
}

.bold{
  font-weight: bold;
}

.inputPassword ~ span {
  grid-area: 3 / 2 / 3 / 2;
}