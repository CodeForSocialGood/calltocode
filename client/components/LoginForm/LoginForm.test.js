/* eslint-env jest */
import React from 'react'
import {shallow} from 'enzyme'

import {LoginForm} from './LoginForm'
let mockProps

beforeEach(() => {
  mockProps = {
    error: null,
    handleSubmit: () => {},
    login: () => {}
  }
})

describe('<LoginForm />', () => {
  describe('error is "Wrong Email"', () => {
    beforeEach(() => { mockProps.error = 'Wrong Email'})

    test('component renders correctly', () => {
      const component = shallow(<LoginForm {...mockProps} />)
      expect(component).toMatchSnapshot()
    })
  })
  describe('error is "Wrong password"', () => {
    beforeEach(() => { mockProps.error = 'Wrong password' })

    test('component renders correctly', () => {
      const component = shallow(<LoginForm {...mockProps} />)
      expect(component).toMatchSnapshot()
    })
  })
})
