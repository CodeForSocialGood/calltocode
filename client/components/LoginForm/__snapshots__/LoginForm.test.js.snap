// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginForm /> error is "Wrong Email" component renders correctly 1`] = `
<form
  className="form"
  id="loginForm"
>
  <h1
    className="title"
  >
    Login
  </h1>
  <Field
    component={[Function]}
    name="email"
  />
  <Field
    component={[Function]}
    name="password"
  />
  <button
    className="buttonSubmit"
    type="submit"
  >
    Submit
  </button>
  <div
    className="errorContent"
  >
    Wrong Email
  </div>
  <Link
    className="forgotPassword"
    key="forgotPassword"
    replace={false}
    to="/forgot-password"
  >
    Forgot Password?
  </Link>
</form>
`;

exports[`<LoginForm /> error is "Wrong password" component renders correctly 1`] = `
<form
  className="form"
  id="loginForm"
>
  <h1
    className="title"
  >
    Login
  </h1>
  <Field
    component={[Function]}
    name="email"
  />
  <Field
    component={[Function]}
    name="password"
  />
  <button
    className="buttonSubmit"
    type="submit"
  >
    Submit
  </button>
  <div
    className="errorContent"
  >
    Wrong password
  </div>
  <Link
    className="forgotPassword"
    key="forgotPassword"
    replace={false}
    to="/forgot-password"
  >
    Forgot Password?
  </Link>
</form>
`;
